### PR TITLE
Fix container rehydration issues and add more E2E tests for them.

### PR DESCRIPTION
**Pull request by jatgarg**
1.) If the data store is loaded after the container is attached, then attach the channel right away. Otherwise add it to local channel context queue, so that it can be attached later with the data store.
2.) Don't set the connection state in local channel context if it is not loaded.
3.) Add more E2E tests.
4.) Clone the snapshot before sanitizing that needs to be passed to dds.

### CODE REVIEW

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 1]
```diff
@@ -212,6 +212,14 @@ export class FluidDataStoreRuntime extends EventEmitter implements IFluidDataSto
                         (content, localOpMetadata) => this.submitChannelOp(path, content, localOpMetadata),
                         (address: string) => this.setChannelDirty(address),
                         tree.trees[path]);
+                    // If the data store is loaded after the container is attached, then attach the channel right away.
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:215]: It's a bit hard to follow code (sorry I missed last changes, so feedback about existing code).

IFluidDataStoreContextType - I feel like this should be removed. IFluidDataStoreContext is casted to it on line 96, saying that isLocalDataStore should be on IFluidDataStoreContext and in current form it should be of boolean | undefined type. It would be great if it can be just boolean.

Could you please promote IF statement on line 202 above enumeration of Object.keys(tree..tree)? dataStoreContext.isLocalDataStore is constant for the loop, so it should be above it.

The code on line 203 - when do we have storage representation for local data store?
This only for Drafts de-serializaiton, right?
Would be great to add a comment on that.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 2]
```diff
@@ -212,6 +212,14 @@ export class FluidDataStoreRuntime extends EventEmitter implements IFluidDataSto
...
+                    // Otherwise add it to local channel context queue, so that it can be attached later with the data
+                    // store.
+                    if (dataStoreContext.attachState !== AttachState.Detached) {
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:218]: It took me a while to understand why it's right. Not sure if we can improve comments, but please consider that.
The key here is that we serialize & de-serialize only bound data stores (which we might need to change, not sure - see Issue #3647).
So context in such case will get attached when container is attached.
Might be worth adding a bit more comments here about it (will be useful as we evaluate above mentioned issue)

Should "attach()" be renamed to "markAttached()" or something like that?
attach() IMHO has a lot of connation to attachGraph, which is action, while here there is no action, just mark up of object.

File: packages/runtime/datastore/src/localChannelContext.ts [hunk 3]
```diff
@@ -4,6 +4,7 @@
  */
 
 import { strict as assert } from "assert";
+import { cloneDeep as clone } from "lodash";
```
Comment (vladsud) [inline packages/runtime/datastore/src/localChannelContext.ts:7]: any reason not to have simple import { cloneDeep } ?

File: packages/runtime/datastore/src/localChannelContext.ts [hunk 4]
```diff
@@ -92,8 +94,8 @@ export class LocalChannelContext implements IChannelContext {
     }
 
     public setConnectionState(connected: boolean, clientId?: string) {
-        // Connection events are ignored if the data store is not yet attached
-        if (!this.attached) {
+        // Connection events are ignored if the data store is not yet attached or loaded
+        if (!(this.attached && this.isLoaded)) {
```
Comment (vladsud) [inline packages/runtime/datastore/src/localChannelContext.ts:98]: I think code is more readable if it's in the form of 
```
if (this.attached && this.isLoaded) {
   this.services.value.deltaConnection.setConnectionState(connected);
}
```
early returns are great if there a lot of nested IFs or code block is long, but for one-liner - not so much

File: packages/runtime/datastore/src/localChannelContext.ts [hunk 5]
```diff
@@ -156,10 +159,6 @@ export class LocalChannelContext implements IChannelContext {
         this.channel = channel;
         this._isLoaded = true;
```
Comment (vladsud) [inline packages/runtime/datastore/src/localChannelContext.ts:160]: Since you are touching this file, any chance you can remove this_isLoaded variable in favor of this implementaiton:
```
    public get isLoaded(): boolean {
        return this.channel !== undefined;
    }
```

File: packages/runtime/datastore/src/localChannelContext.ts [hunk 6]
```diff
@@ -156,10 +159,6 @@ export class LocalChannelContext implements IChannelContext {
...
 
-        if (this.attached) {
```
Comment (vladsud) [inline packages/runtime/datastore/src/localChannelContext.ts:159]: Can you please let me know why this code is not needed any more? Can't figure it out

**Review by vladsud — APPROVED** [review]

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:218]: I'm confused. i thought only the container's attached to storage, and everything else was bound to it's parent

---
In reply to: [489152975](https://github.com/microsoft/FluidFramework/pull/3645#discussion_r489152975) [](ancestors = 489152975)

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 1 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:215]: Added a comment. Tony suggested that we don't want to expose it IFluidDataStoreContext and just want to keep it internal to runtime for implementation and I agree. So the typecast is necessary and we know that the boolean will be present(not undefined) and we assign it in dataStoreContext.ts.

File: packages/runtime/datastore/src/localChannelContext.ts (hunk 3 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/localChannelContext.ts:7]: Changed it

File: packages/runtime/datastore/src/localChannelContext.ts (hunk 4 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/localChannelContext.ts:98]: Changed

File: packages/runtime/datastore/src/localChannelContext.ts (hunk 5 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/localChannelContext.ts:160]: Changed

File: packages/runtime/datastore/src/localChannelContext.ts (hunk 6 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/localChannelContext.ts:159]: Actually this was redundant because during load we call factory.load which actually assign services and attach delta handler. So we don't need to do it again. I added the comment on factory.load about it in the PR.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 2 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:218]: @anthony-murphy Yes you are right. When the containers attach to storage, everything bound to it gets attached to storage. 
In this case also, it is actually present on the storage and therefore already attached but due to delay loading of the data store, channel was not mark attached as vlad said in his comment. So here we are just marking it up.

Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:218]: So I think I will take vlad suggestion and rename the api.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: This to me does not make sense. These are two different layers - container layer, and data store layer. I, as a developer, need to be able to implement my own version of DataStoreRuntime, using public API surface. Making comment that it's not piblic is equivalent to me that I can't do this work, without reverse-engineering and using undocumented API.

Maybe it needs to take another form, but data store runtime layer should not rely on anything non-public in container layer.

Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: But IFluidDataStoreContextType is also public it is just that it is not in main set of apis in IFluidDataStoreContext as it sounds more internal to runtime implementation. Do you suggest adding the property in IFluidDataStoreContext interface?
@vladsud @anthony-murphy ?

Comment (vladsud) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: The contract should be clear. I do not see an advantage of existing behavior compared to putting this property on main object as boolean | undefined. Next improvement step if we can drop undefined part.

Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: It is boolean. How is it undefined?

Comment (vladsud) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: It's undefined on IFluidDataStoreContext interface. At least that's how I'd interpret this code that cases and object without property to derive interface that has this property. I'd rather see it that way (undefined | boolean) on original interface (and no additional interfaces) if that's the intention, or better - as boolean property on IFluidDataStoreContext
